edb/server/protocol/auth_ext/pkce.py

```diff
@@ -0,0 +1,94 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2023-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import json
+
+from edb.server.protocol import execute
+from . import data
+
+
+class PKCE:
+    def __init__(self, db):
+        self.db = db
+
+    async def create(self, challenge: str):
```

Comment:
IMO all of these methods should be top level functions in the `pkce.py` module. There's no need for the `PKCE` class as it holds no state; functions can just accept `db` as the first argument.

Reply:
And then I'd move the PKCE dataclass definition to this file; it would make much more sense.

Reply:
Yeah, this would be my natural style in TypeScript, so makes sense!